models/icepack/model_mod.f90

```diff
@@ -0,0 +1,885 @@
+integer(i8),         intent(in)  :: offset
+real(r8),            intent(in)  :: lon, lat
+integer,             intent(in)  :: var_type
+integer,             intent(in)  :: cat_signal
+real(r8),            intent(out) :: expected_obs(ens_size)
+integer,             intent(out) :: istatus(ens_size)
+
+real(r8) :: work_expected_obs(ens_size)
+integer  :: e, iterations, Niterations
+integer(i8) :: state_index
+
+if ( .not. module_initialized ) call static_init_model
+
+istatus = 0
+if (var_type == 14) then
+  e = 1
+else if (var_type == 15) then
+  e = 2
+else if (var_type == 16) then
+  e = 3
+endif
+if ( cat_signal < 1 )  then
+   Niterations = Ncat ! only iterate if aggregating over all types
+else
+   Niterations = 1 ! no need to iterate
+endif
+work_expected_obs = 0.0_r8
+expected_obs = 0.0_r8
+do iterations = 1, Niterations
+
+   ! FIXME: this should use the state structure routine 'get_dart_vector_index'
+   ! to get the start of the next category layer.  this code assumes it knows
+   ! exactly how the state vector is laid out (reasonable, but might not be true
+   ! in future versions of dart.)
+   !next_offset = offset + (iterations-1)*Nx 
+   !print*,'offset',offset
+   state_index = get_dart_vector_index(grid_oi,int(offset,i4),1, domain_id, e)
+   work_expected_obs = get_state(state_index,state_handle)
+   expected_obs = expected_obs+work_expected_obs
+enddo
+end subroutine lon_lat_interpolate
+
+!------------------------------------------------------------------
+! Given an integer index into the state vector structure, returns 
+! the associated location.
+
+subroutine get_state_meta_data(index_in, location, var_type)
+
+integer(i8),         intent(in)  :: index_in
+type(location_type), intent(out) :: location
+integer,             intent(out), optional :: var_type
+
+real(r8) :: lat, lon, rcat
+integer  :: ni_index, hold_index, cat_index, local_var, var_id
+! these should be set to the actual location and state quantity
+
+if ( .not. module_initialized ) call static_init_model
+
+call get_model_variable_indices(index_in, ni_index, cat_index, hold_index, var_id=var_id)
+call get_state_kind(var_id, local_var)
+
+lon = TLON(ni_index)
+lat = TLAT(ni_index)
+
+if (debug > 5) print *, 'lon, lat, cat_index = ', lon, lat, cat_index
+rcat     = cat_index*1.0_r8
+location = set_location(lon, lat, rcat, VERTISLEVEL)
+
+if (present(var_type)) then
+   var_type = local_var
+endif
+
+end subroutine get_state_meta_data
+
+!------------------------------------------------------------------
+! Given an integer index into the state vector structure, returns the kind,
+! and both the starting offset for this kind, as well as the offset into
+! the block of this kind.
+
+subroutine get_state_kind(var_ind, var_type)
+ integer, intent(in)  :: var_ind
+ integer, intent(out) :: var_type
+
+if ( .not. module_initialized ) call static_init_model
+
+var_type = state_kinds_list(var_ind)
+
+end subroutine get_state_kind
+
```

Comment:
Is there are reason to use a local version rather than the state_structure routines?